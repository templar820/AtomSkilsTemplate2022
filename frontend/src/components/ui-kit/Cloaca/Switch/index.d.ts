import * as React from 'react';
import { EmotionStylesType } from 'lib-ui';

export interface SwitchProps {
    /** State of the switch. In uncontrolled mode defines initial state */
    checked?: boolean;
    /** State of the component can't be changed by user if disabled is true */
    disabled?: boolean;
    /** Add loader atop of the component. Also switches component do disabled */
    isLoading?: boolean;
    /** Defines whether component checking controlled by externally or by itself */
    isControlled?: boolean;
    /** Emotion styles applied to the wrapper of component. Overrides default. */
    wrapperStyle?: EmotionStylesType;
    /** Overrides thumb styles. Receives emotion object or function, which handle component's props and returns emotion */
    thumbStyle?: EmotionStylesType;
    /** Overrides box styles. Receives emotion object or function, which handle component's props and returns emotion */
    boxStyle?: EmotionStylesType;
    /** Overrides props passed to loader component (look lib-ui/utils/LoaderComponent for details) */
    loaderProps?: object;
    /** Overrides default color scheme object */
    colorSchemeExtra?: object;
    /** Overrides default size based config  */
    switchSizesExtra?: object;
    /** Defines which config will be applied to the switch */
    elmSize?: 'sm' | 'md';
    /** Defines what icons (mentioned in config) will be shown on both sides of the Switch till not overwritten. No icons if undefined */
    iconsType?: 'numeric' | 'iconic';
    /** Defines what icon (or any sign) will be shown when switch is checked. if not set, icon will be chosen by iconsType */
    checkedIcon?: React.ReactNode;
    /** Defines what icon (or any sign) will be shown when switch is not checked. if not set, icon will be chosen by iconsType */
    uncheckedIcon?: React.ReactNode;
    /** Defines how long checking animation lasts. In ms */
    animationDuration?: number;
    /** Callback fires on check change, func(newCheckedState) */
    onCheckChange?: (newCheckedState: boolean) => void;
}

declare const Switch: React.ComponentType<SwitchProps>;

export default Switch;
