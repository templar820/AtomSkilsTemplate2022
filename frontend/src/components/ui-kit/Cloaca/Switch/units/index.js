import React from 'react';
import styled from '@emotion/styled';

import Box from './Box';
import Icon from './Icon';
import Loader from './Loader';
import applyEmotionStyle from 'lib-root/utils/applyEmotionStyle';

const StyledWrapper = styled.label`
    display: inline-flex;
    margin: 0;
    position: relative;
    pointer-events: ${({ isLoading, disabled }) => (isLoading || disabled) && 'none'};
    ${({ wrapperStyle, ...restProps }) => applyEmotionStyle(wrapperStyle, restProps)};
`;

const StyledInput = styled.input`
    opacity: 0;
`;

export { Icon, Box, Loader, StyledInput, StyledWrapper };
