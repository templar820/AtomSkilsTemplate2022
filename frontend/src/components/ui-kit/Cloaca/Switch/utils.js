import { hexToRGBA } from 'lib-root/utils/styleMixins';

const switchSizes = {
    md: {
        wrapperWidth: '62px',
        wrapperHeight: '32px',
        thumbWidth: '30px',
        fontSize: '15px',
        defaultIconSize: '20px',
        loaderScale: '0.5'
    },
    sm: {
        wrapperWidth: '42px',
        wrapperHeight: '22px',
        thumbWidth: '20px',
        fontSize: '10px',
        defaultIconSize: '16px',
        loaderScale: '0.4'
    }
};

const getColorScheme = (colors) => ({
    main: hexToRGBA(colors.primary, 0.9),
    border: colors.primary,
    additional: colors.GrayScale_0,
    inactive: colors.GrayScale_50,
    inactiveAccent: colors.GrayScale_200,
    inactiveBorder: colors.GrayScale_100,
    thumbShadow: hexToRGBA(colors.GrayScale_700, 0.25)
});

const iconConfig = {
    checked: {
        position: 'left',
        numeric: '1',
        iconName: 'Success'
    },
    unchecked: {
        position: 'right',
        numeric: '0',
        iconName: 'Cross_exit'
    }
};

function handleKeyUp(key, cb) {
    if (key === 'ArrowLeft') cb(false);
    if (key === 'ArrowRight') cb(true);
}

export { switchSizes, getColorScheme, iconConfig, handleKeyUp };
