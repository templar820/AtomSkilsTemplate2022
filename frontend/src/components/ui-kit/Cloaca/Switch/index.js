import React, { useState, useContext } from 'react';
import PropTypes from 'prop-types';

import { ColorsContext } from 'lib-root/colors';

import { Box, Icon, Loader, StyledInput, StyledWrapper } from './units';

import { getColorScheme, switchSizes as switchSizesDefault, handleKeyUp } from './utils';
import { setClassName } from 'lib-root/utils/styleMixins';

/**
 * Переключатель
 */
const Switch = React.forwardRef(
    (
        {
            className,
            isControlled,
            colorSchemeExtra,
            switchSizesExtra,
            onCheckChange,
            checkedIcon,
            uncheckedIcon,
            isLoading,
            iconsType,
            elmSize,
            animationDuration,
            disabled: disabledByProps,
            checked: checkedByProps,
            wrapperStyle,
            thumbStyle,
            boxStyle,
            loaderProps,
            ...restProps
        },
        ref
    ) => {
        const colors = useContext(ColorsContext);

        // Init hooks
        const [focused, setFocused] = useState(false);
        const [controlledChecked, setControlledChecked] = useState(checkedByProps);

        // Deciding how to handle checking
        const handleCheck = (newCheckedState) => {
            if (checked === newCheckedState) return;
            if (!isControlled) {
                setControlledChecked(newCheckedState);
            }
            if (onCheckChange) onCheckChange(newCheckedState);
        };

        // Calculating further props
        const checked = isControlled ? checkedByProps : controlledChecked;
        const disabled = isLoading ? true : disabledByProps;

        // Extending configs and choosing suiting ones
        const colorScheme = { ...getColorScheme(colors), ...colorSchemeExtra };
        const currentSwitchSizes = { ...switchSizesDefault, ...switchSizesExtra }[elmSize];

        return (
            <StyledWrapper
                {...{ isLoading, disabled, wrapperStyle }}
                className={setClassName({ props: { className }, name: 'switch' })}>
                <Box
                    {...{
                        checked,
                        isLoading,
                        disabled,
                        colorScheme,
                        currentSwitchSizes,
                        focused,
                        animationDuration,
                        thumbStyle,
                        boxStyle
                    }}>
                    <Icon {...{ checked, currentSwitchSizes, checkedIcon, uncheckedIcon, iconsType }} />
                    <StyledInput
                        onFocus={() => setFocused(true)}
                        onBlur={() => setFocused(false)}
                        onChange={(e) => handleCheck(e.target.checked)}
                        onKeyUp={(e) => handleKeyUp(e.key, handleCheck)}
                        type="checkbox"
                        {...{ checked, ref, ...restProps }}
                    />
                </Box>
                <Loader {...{ currentSwitchSizes, isLoading, loaderProps }} />
            </StyledWrapper>
        );
    }
);

Switch.displayName = 'Switch';
Switch.propTypes = {
    /** State of the switch. In uncontrolled mode defines initial state */
    checked: PropTypes.bool,
    /** State of the component can't be changed by user if disabled is true */
    disabled: PropTypes.bool,
    /** Add loader atop of the component. Also switches component do disabled */
    isLoading: PropTypes.bool,
    /** Defines whether component checking controlled by externally or by itself */
    isControlled: PropTypes.bool,
    /** Emotion styles applied to the wrapper of component. Overrides default. */
    wrapperStyle: PropTypes.oneOfType([PropTypes.func, PropTypes.object, PropTypes.string]),
    /** Overrides thumb styles. Receives emotion object or function, which handle component's props and returns emotion */
    thumbStyle: PropTypes.oneOfType([PropTypes.func, PropTypes.object, PropTypes.string]),
    /** Overrides box styles. Receives emotion object or function, which handle component's props and returns emotion */
    boxStyle: PropTypes.oneOfType([PropTypes.func, PropTypes.object, PropTypes.string]),
    /** Overrides props passed to loader component (look lib-ui/utils/LoaderComponent for details) */
    loaderProps: PropTypes.object,
    /** Overrides default color scheme object */
    colorSchemeExtra: PropTypes.object,
    /** Overrides default size based config  */
    switchSizesExtra: PropTypes.object,
    /** Defines which config will be applied to the switch */
    elmSize: PropTypes.oneOf(['sm', 'md']),
    /** Defines what icons (mentioned in config) will be shown on both sides of the Switch till not overwritten. No icons if undefined */
    iconsType: PropTypes.oneOf(['numeric', 'iconic']),
    /** Defines what icon (or any sign) will be shown when switch is checked. if not set, icon will be chosen by iconsType */
    checkedIcon: PropTypes.node,
    /** Defines what icon (or any sign) will be shown when switch is not checked. if not set, icon will be chosen by iconsType */
    uncheckedIcon: PropTypes.node,
    /** Defines how long checking animation lasts. In ms */
    animationDuration: PropTypes.number,
    /** Callback fires on check change, func(newCheckedState) */
    onCheckChange: PropTypes.func
};
Switch.defaultProps = {
    checked: true,
    disabled: false,
    isLoading: false,
    isControlled: true,
    elmSize: 'md',
    iconsType: undefined,
    colorSchemeExtra: {},
    switchSizesExtra: {},
    loaderProps: {},
    animationDuration: 300
};

export default Switch;
