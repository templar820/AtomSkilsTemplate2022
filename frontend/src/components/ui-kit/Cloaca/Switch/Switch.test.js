import React from 'react';

import serializer from 'jest-emotion';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import Switch from './';

expect.addSnapshotSerializer(serializer);

Enzyme.configure({ adapter: new Adapter() });

const mockFn = jest.fn();

const colorSc = {
    main: 'blue',
    inactive: 'red',
    border: 'black',
    inactiveBorder: 'black',
    additional: 'lightgrey'
};

const checkedIconProp = <div id="checkedIco">Styled ico</div>;
const uncheckedIconProp = <div id="uncheckedIco">Styled ico</div>;

const props = {
    disabled: false,
    checked: false,
    isControlled: true,
    checkedIcon: checkedIconProp,
    uncheckedIcon: uncheckedIconProp,
    iconsType: 'iconic',
    onCheckChange: mockFn
};

describe('<Switch />', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = mount(<Switch {...props} />);
    });

    afterEach(() => {
        wrapper.unmount();
    });

    test('should match snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('should should render correctly', () => {
        expect(typeof wrapper.props().onCheckChange).toBe('function');

        expect(wrapper.find('svg')).toBeTruthy();

        expect(wrapper.props()).toHaveProperty('checked', props.checked);
        expect(wrapper.props()).toHaveProperty('disabled', props.disabled);
        expect(wrapper.props()).toHaveProperty('isControlled', props.isControlled);
        expect(wrapper.props()).toHaveProperty('wrapperStyle', props.wrapperStyle);
        expect(wrapper.props()).toHaveProperty('thumbStyle', props.thumbStyle);
        expect(wrapper.props()).toHaveProperty('boxStyle', props.boxStyle);
        expect(wrapper.props()).toHaveProperty('iconsType', props.iconsType);
        expect(wrapper.props()).toHaveProperty('checkedIcon', props.checkedIcon);
        expect(wrapper.props()).toHaveProperty('uncheckedIcon', props.uncheckedIcon);
        expect(wrapper.props()).toHaveProperty('onCheckChange', props.onCheckChange);
    });

    test('should input checked', () => {
        expect(wrapper.find('input')).not.toBeChecked();
        wrapper.setProps({ checked: true });
        expect(wrapper.find('input')).toBeChecked();
    });

    test('should change have another colors', () => {
        expect(wrapper.prop('colorSchemeExtra')).toEqual({});
        wrapper.setProps({ colorSchemeExtra: colorSc });
        expect(wrapper.prop('colorSchemeExtra')).not.toEqual({});
    });

    test('should have checkedIcon', () => {
        expect(wrapper.find('div[id="checkedIco"]')).toBeTruthy();
        expect(wrapper.find('div[id="uncheckedIco"]')).toBeTruthy();
    });

    test('should call onCheckChange function', () => {
        wrapper.find('input').simulate('change', { target: { value: true } });
        expect(mockFn).toHaveBeenCalled();
    });

    test('should have numeric ', () => {
        wrapper.setProps({ iconsType: 'numeric' });
        wrapper.update();
        expect(wrapper.find('span>span')).toBeTruthy();
    });
});
