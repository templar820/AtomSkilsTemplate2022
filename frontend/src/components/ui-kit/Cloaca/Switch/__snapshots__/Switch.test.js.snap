// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> should match snapshot 1`] = `
.emotion-6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin: 0;
  position: relative;
}

.emotion-4 {
  display: inline-block;
  width: 62px;
  height: 32px;
  font-size: 15px;
  position: relative;
  border-radius: 100px;
  -webkit-transition-property: background-color,border-color,color;
  transition-property: background-color,border-color,color;
  -webkit-transition-duration: 300ms;
  transition-duration: 300ms;
  border: 1px solid #EDEDED;
  background-color: #fafafa;
  color: #D8D8D8;
  cursor: pointer;
}

.emotion-4:before {
  content: '';
  width: 30px;
  height: 30px;
  background-color: #ffffff;
  border-radius: 50%;
  position: absolute;
  top: 0;
  left: 0;
  -webkit-transform: translate(0);
  -ms-transform: translate(0);
  transform: translate(0);
  box-shadow: 0 0 4px rgba(51,51,51,0.25);
  -webkit-transition: -webkit-transform 300ms;
  -webkit-transition: transform 300ms;
  transition: transform 300ms;
}

.emotion-0 {
  position: absolute;
  top: 50%;
  right: 25%;
  -webkit-transform: translate(50%,-50%);
  -ms-transform: translate(50%,-50%);
  transform: translate(50%,-50%);
}

.emotion-2 {
  opacity: 0;
}

<Switch
  animationDuration={300}
  checked={false}
  checkedIcon={
    <div
      id="checkedIco"
    >
      Styled ico
    </div>
  }
  colorSchemeExtra={Object {}}
  disabled={false}
  elmSize="md"
  iconsType="iconic"
  isControlled={true}
  isLoading={false}
  loaderProps={Object {}}
  onCheckChange={[MockFunction]}
  switchSizesExtra={Object {}}
  uncheckedIcon={
    <div
      id="uncheckedIco"
    >
      Styled ico
    </div>
  }
>
  <StyledWrapper
    className="CLR-switch"
    disabled={false}
    isLoading={false}
  >
    <label
      className="CLR-switch emotion-6 emotion-7"
      disabled={false}
    >
      <_default
        animationDuration={300}
        checked={false}
        colorScheme={
          Object {
            "additional": "#ffffff",
            "border": "#00B956",
            "inactive": "#fafafa",
            "inactiveAccent": "#D8D8D8",
            "inactiveBorder": "#EDEDED",
            "main": "rgba(0, 185, 86, 0.9)",
            "thumbShadow": "rgba(51, 51, 51, 0.25)",
          }
        }
        currentSwitchSizes={
          Object {
            "defaultIconSize": "20px",
            "fontSize": "15px",
            "loaderScale": "0.5",
            "thumbWidth": "30px",
            "wrapperHeight": "32px",
            "wrapperWidth": "62px",
          }
        }
        disabled={false}
        focused={false}
        isLoading={false}
      >
        <StyledBox
          animationDuration={300}
          checked={false}
          colorScheme={
            Object {
              "additional": "#ffffff",
              "border": "#00B956",
              "inactive": "#fafafa",
              "inactiveAccent": "#D8D8D8",
              "inactiveBorder": "#EDEDED",
              "main": "rgba(0, 185, 86, 0.9)",
              "thumbShadow": "rgba(51, 51, 51, 0.25)",
            }
          }
          currentSwitchSizes={
            Object {
              "defaultIconSize": "20px",
              "fontSize": "15px",
              "loaderScale": "0.5",
              "thumbWidth": "30px",
              "wrapperHeight": "32px",
              "wrapperWidth": "62px",
            }
          }
          disabled={false}
          finalthumbStyle={
            Object {
              "map": undefined,
              "name": "1ts1wle",
              "next": undefined,
              "styles": "content:'';width:30px;height:30px;background-color:#ffffff;border-radius:50%;position:absolute;top:0;left:0;transform:translate(0);box-shadow:0 0 4px rgba(51, 51, 51, 0.25);transition:transform 300ms;;",
              "toString": [Function],
            }
          }
          focused={false}
          isLoading={false}
        >
          <span
            checked={false}
            className="emotion-4 emotion-5"
            disabled={false}
          >
            <_default
              checked={false}
              checkedIcon={
                <div
                  id="checkedIco"
                >
                  Styled ico
                </div>
              }
              currentSwitchSizes={
                Object {
                  "defaultIconSize": "20px",
                  "fontSize": "15px",
                  "loaderScale": "0.5",
                  "thumbWidth": "30px",
                  "wrapperHeight": "32px",
                  "wrapperWidth": "62px",
                }
              }
              iconsType="iconic"
              uncheckedIcon={
                <div
                  id="uncheckedIco"
                >
                  Styled ico
                </div>
              }
            >
              <StyledIcon
                position="right"
              >
                <span
                  className="emotion-0 emotion-1"
                >
                  <div
                    id="uncheckedIco"
                  >
                    Styled ico
                  </div>
                </span>
              </StyledIcon>
            </_default>
            <StyledInput
              checked={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyUp={[Function]}
              type="checkbox"
            >
              <input
                checked={false}
                className="emotion-2 emotion-3"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyUp={[Function]}
                type="checkbox"
              />
            </StyledInput>
          </span>
        </StyledBox>
      </_default>
      <_default
        currentSwitchSizes={
          Object {
            "defaultIconSize": "20px",
            "fontSize": "15px",
            "loaderScale": "0.5",
            "thumbWidth": "30px",
            "wrapperHeight": "32px",
            "wrapperWidth": "62px",
          }
        }
        isLoading={false}
        loaderProps={Object {}}
      />
    </label>
  </StyledWrapper>
</Switch>
`;
